core/serialize/serialize.cc

```diff
@@ -1035,19 +981,90 @@ void SerializerImpl::unpickleGS(UnPickler &p, GlobalState &result) {
         Timer timeit(result.tracer(), "moving");
         result.fileRefByPath = std::move(fileRefByPath);
         result.files = std::move(files);
-        result.utf8Names = std::move(utf8Names);
-        result.constantNames = std::move(constantNames);
-        result.uniqueNames = std::move(uniqueNames);
         result.classAndModules = std::move(classAndModules);
         result.methods = std::move(methods);
         result.fields = std::move(fields);
         result.typeArguments = std::move(typeArguments);
         result.typeMembers = std::move(typeMembers);
-        result.namesByHash = std::move(namesByHash);
     }
     result.sanityCheck();
 }
 
+void SerializerImpl::pickleNameTable(Pickler &p, const GlobalState &gs) {
+    p.putU4(gs.utf8Names.size());
+    for (const auto &n : gs.utf8Names) {
+        pickle(p, n);
+    }
+    p.putU4(gs.constantNames.size());
+    for (const auto &n : gs.constantNames) {
+        pickle(p, n);
+    }
+    p.putU4(gs.uniqueNames.size());
+    for (const auto &n : gs.uniqueNames) {
+        pickle(p, n);
+    }
+
+    p.putU4(gs.namesByHash.size());
+    for (const auto &s : gs.namesByHash) {
+        p.putU4(s.hash);
+        p.putU4(s.rawId);
+    }
+}
+
+void SerializerImpl::unpickleNameTable(UnPickler &p, GlobalState &result) {
```

Comment:
We might want to do something here that asserts that the thing we're unpickling matches whatever is there?

For example, something like make sure that a handful of well-known names still have the same IDs in the new table" or "there are at least as many names in the deserialized data as there were in the current `result` or "for a random sample of like 10 NameRef IDs, the raw strings they represent are the same in the old and new GlobalState"

Reply:
That's a good point, it would be nice to have a canary for any cache issues.